types/xdg_shell/wlr_xdg_surface.c

```diff
@@ -346,19 +345,65 @@ static void xdg_surface_handle_surface_commit(struct wl_listener *listener,
 	}
 }
 
+static void xdg_surface_handle_surface_cache(struct wl_listener *listener,
+		void *data) {
+	struct wlr_xdg_surface *surface =
+		wl_container_of(listener, surface, surface_cache);
+
+	if (!surface->pending.has_geometry) {
```

Comment:
I think we should still cache the state here:

- makes no sense if we have other properties too (fullscreen, tiled, etc.)
- some clients do not use window geometry, and leave it to the default of 'bounding box of surface+subsurfaces'. We'd still want to track configure serials for such surfaces, and for that we need to iterate the cached state of the xdg_surface.

Reply:
> makes no sense if we have other properties too (fullscreen, tiled, etc.)

Fullscreen/tiled/etc are not part of the double-buffered toplevel state.

> some clients do not use window geometry, and leave it to the default of 'bounding box of surface+subsurfaces'. We'd still want to track configure serials for such surfaces, and for that we need to iterate the cached state of the xdg_surface.

The configure serials are missing from the cached xdg_surface state. We should add it, and only cache the state if the pending geometry or pending configure serial has changed.